### DISCUSSION TITLE
Colors issues with ILI9341_2 on ESP32-2432S028R with 2 USB ports

### DISCUSSION THREAD

**Discussion by TheNitek** [Bodmer/TFT_eSPI#3018] [Compatible displays and setup files]
I am using bodmer/TFT_eSPI@^2.5.33 on a ESP32-2432S028R with 2 USB ports, which seems to have another variant of the ILI9341_2 (compared to the 1 USB variant).
In general it works well (just needs TFT_INVERSION_ON)

Only issue is that the colors are way of, which looks like very poor gamma settings:

![PXL_20231118_134626542](https://github.com/Bodmer/TFT_eSPI/assets/6240544/e195c3bc-8f69-4ce5-b907-09d7c08a3181)

While I don't have a clue how gamma works on the ILI9341, I just started experimenting and noticed that this highly improves the picture quality:

```
  tft.writecommand(ILI9341_GAMMASET); //Gamma curve selected
  tft.writedata(2);
  delay(120);
  tft.writecommand(ILI9341_GAMMASET); //Gamma curve selected
  tft.writedata(1);
```

![PXL_20231118_145142318](https://github.com/Bodmer/TFT_eSPI/assets/6240544/e927e3fc-5612-486f-b8db-5a076750b4bb)

Does the gamma table maybe need a change because of the inversion?

**Comment by harrytrinh9** [comment]
I have same issue with @TheNitek 
The image display actually very poor quality after call inverted display function.
![20231125_070135](https://github.com/Bodmer/TFT_eSPI/assets/90092107/655ab195-4db1-4891-847e-1e3b63ad8e5b)

**Comment by zehn-cat** [comment]
I had the same problem with the same board. I get normal colors by commenting out these two lines from ILI9341_Init.h:
writecommand(0xE0); //Set Gamma
writecommand(0XE1); //Set Gamma

**Comment by harrytrinh9** [comment]
Thanks @zehn-cat so much !
Your solution is work. But the gamma still not pretty
![20231126_191128](https://github.com/Bodmer/TFT_eSPI/assets/90092107/e35325f6-9598-4cbb-971e-2ac99ff63e79)

**Reply by Benedito821** [reply]
Hello @harrytrinh9 ,
Could you please share your PNG array file (.h file) so that I could test on my board?

**Comment by hello-world-dot-c** [comment]
@TheNitek This works great for me, thanks! How did you come up with this sequence? Did you find it somewhere?

**Comment by Bodmer** [comment]
Unfortunately the TFT screens vary in supplier and this screen appears to need different settings. Also some screens appear to have cloned chips fitted that are not OEM.

So some tweaking may be needed in these cases.

**Reply by TheNitek** [reply]
Is there a way to have different default predefinied gamma curves for easy switching? Display works fine otherways, just the colors looks bad because of the gamma this lib applies by default

**Reply by Bodmer** [reply]
All Gamma settings are coded into the init sequence, for example here:
https://github.com/Bodmer/TFT_eSPI/blob/master/TFT_Drivers/ILI9341_Init.h
If you create a new init set for that display I can added it. The existing settings work fine one my displays but there are many vendors. Ideally use the settings supplied form the vendor but with cheap displays there is very little supplier support...

**Reply by TheNitek** [reply]
Yeah, I couldn't find anything usefull either, but simply NOT setting any gamma values seemed to work fine. So if I wanted to provide a pull request, I'd just add a #ifdef to this file?

**Reply by dart1280** [reply]
You have seen these haven't you - 
// If colours are inverted (white shows as black) then uncomment one of the next
// 2 lines try both options, one of the options should correct the inversion.

// #define TFT_INVERSION_ON
// #define TFT_INVERSION_OFF

and - 

// For ST7789 and ILI9341 ONLY, define the colour order IF the blue and red are swapped on your display
// Try ONE option at a time to find the correct colour order for your display

//  #define TFT_RGB_ORDER TFT_RGB  // Colour order Red-Green-Blue
//  #define TFT_RGB_ORDER TFT_BGR  // Colour order Blue-Green-Red

both from SetupX_Template.h

**Reply by CoolmdXi** [reply]
I think Bodmer may have seen these as he wrote the libraies LMAO

**Comment by hai8825** [comment]
I am having similar issue with the MakerFocus 1.3inch TFT LCD Display Module 3.3V with SPI Interface ST7789 IC Driver display. Wish I could know what I need to change to fix the gamma and the colors on it for my project. I have been commenting stuff out from ST7789_init.h but no luck so far.

**Reply by dart1280** [reply]
I added #define TFT_INVERSION_ON and #define TFT_RGB_ORDER TFT_BGR to the displays Setup file in .............\Arduino\libraries\TFT_eSPI\User_Setups folder, It wasn't really a gamma issue because my display was just showing the wrong colours but it fixed my ESP32-2432S028R display.

```
#define ILI9341_2_DRIVER     

#define TFT_WIDTH  240
#define TFT_HEIGHT 320
#define TFT_BL   21            
#define TFT_BACKLIGHT_ON HIGH  
#define TFT_MOSI 13 // In some display driver board, it might be written as "SDA" and so on.
#define TFT_SCLK 14
#define TFT_CS   15  // Chip select control pin
#define TFT_DC   2  // Data Command control pin
#define TFT_RST  12  // Reset pin (could connect to Arduino RESET pin)
#define TFT_BL   21  // LED back-light
//#define TFT_RGB_ORDER TFT_RGB  // Colour order Red-Green-Blue
#define TFT_RGB_ORDER TFT_BGR  // Colour order Blue-Green-Red

#define TFT_INVERSION_ON   //if the colours are wrong try the other setting
// #define TFT_INVERSION_OFF


//#define TOUCH_CS 33     // Chip select pin (T_CS) of touch screen

#define LOAD_GLCD   // Font 1. Original Adafruit 8 pixel font needs ~1820 bytes in FLASH
#define LOAD_FONT2  // Font 2. Small 16 pixel high font, needs ~3534 bytes in FLASH, 96 characters
#define LOAD_FONT4  // Font 4. Medium 26 pixel high font, needs ~5848 bytes in FLASH, 96 characters
#define LOAD_FONT6  // Font 6. Large 48 pixel font, needs ~2666 bytes in FLASH, only characters 1234567890:-.apm
#define LOAD_FONT7  // Font 7. 7 segment 48 pixel font, needs ~2438 bytes in FLASH, only characters 1234567890:-.
#define LOAD_FONT8  // Font 8. Large 75 pixel font needs ~3256 bytes in FLASH, only characters 1234567890:-.
//#define LOAD_FONT8N // Font 8. Alternative to Font 8 above, slightly narrower, so 3 digits fit a 160 pixel TFT
#define LOAD_GFXFF  // FreeFonts. Include access to the 48 Adafruit_GFX free fonts FF1 to FF48 and custom fonts

// Comment out the #define below to stop the SPIFFS filing system and smooth font code being loaded
// this will save ~20kbytes of FLASH
#define SMOOTH_FONT


#define SPI_FREQUENCY  55000000 
#define SPI_READ_FREQUENCY  20000000
```

**Reply by hai8825** [reply]
Yeah I have those set too, but the colors are still to white tone, and sometimes looked like crushed. Not sure what settings I can mess with to somehow improve it.

**Comment by TheNitek** [comment]
For the lost souls still landing on this issue: The version with the 2 USB ports has a ST7789

So the "correct" config is
ST7789_DRIVER
TFT_INVERSION_OFF
TFT_RGB_ORDER=TFT_BGR

**Reply by Grey-Lancaster** [reply]
This worked for me. As above mine is usb-c but apears to work with 9341

[env:cyd]
build_flags =
	${env.build_flags}
	-DILI9341_2_DRIVER
	

[env:cyd2usb]
build_flags =
	${env.build_flags}
	-DST7789_DRIVER
	-DTFT_INVERSION_OFF

[env:cyd2b]
build_flags =
	${env.build_flags}
	-DILI9341_2_DRIVER
	-DTFT_INVERSION_ON
	-DENV_CYD2B
	-DUSE_GAMMA_CORRECTION

then in .ino

  // Start the tft display and set it to black
  tft.init();
  tft.setRotation(1); //This is the display in landscape

 // Define the ILI9341_GAMMASET if it is not included in your library
#define ILI9341_GAMMASET 0x26  // Gamma curve command for ILI9341
#ifdef ENV_CYD2B
  // Code specific to cyd2b
    // gamma curve setup 
  tft.writecommand(ILI9341_GAMMASET); // Gamma curve selected
  tft.writedata(2);
  delay(120);
  tft.writecommand(ILI9341_GAMMASET); // Gamma curve selected
  tft.writedata(1);  
#endif


  // Clear the screen before writing to it
  tft.fillScreen(TFT_BLACK);

**Comment by Benedito821** [comment]
Toggling bit BGR of the register _Memory Access Control_ _(0x36)_ to change the colour representation sequence dind't work for me. So I did manually swap the code of the colours in my defines (I'm using STM32) , i.e. red (0xF800) became 0x001F , blue (0x001F) became 0xF800 and so on. Before I call any interface that takes colour as a parameter, I swap the the colour first. So, either we hard-swap it or we swap each time we call an interface. I find the later one better.

**Reply by Benedito821** [reply]
Although when I try to push an image it seems not working that much, and it doesn't seem being a gamma issue.

![20241112_223955](https://github.com/user-attachments/assets/04d49478-e553-4ef7-a8fe-81094294e60e)

![parrot_240x240](https://github.com/user-attachments/assets/12c0be4d-dfc5-4477-a7f0-e975aad8c38c)

**Comment by Grey-Lancaster** [comment]
try the CYD2 alternate on this page.  It assumes you have your pic 320x240 on the sd card
https://grey-lancaster.github.io/CYD-PhotoFrame/
wifi manager mdns photoframe.local

**Comment by lobonz** [comment]
Great thread, helped me a lot first time dealing with a screen, solved the gamma issue, once I messed around with all the driver and color options.
The CYD I have is: TENSTAR ESP32 Development Board with 2.8inch LCD LVGL WIFI&Bluetooth 240*320 Smart Display Screen TFT Module With Touch WROOM (Aliexpress)
Display: ILI9341 - using ILI9341_2_DRIVER, TFT_BGR, TFT_INVERSION_ON

To fix the milky gamma issue I added this to the top of my sketch which sets correct gamma on my CYD ILI9341_2 panel
```
#ifndef ILI9341_GAMMASET
#define ILI9341_GAMMASET 0x26
#endif
